hypothesis-python/src/hypothesis/internal/conjecture/datatree.py

```diff
@@ -17,210 +17,259 @@
-        # buffer contents.
-        self.block_sizes = {}
+    def __init__(self):
+        self.root = TreeNode()
 
     @property
     def is_exhausted(self):
         """Returns True if every possible node is dead and thus the language
         described must have been fully explored."""
-        return 0 in self.dead
-
-    def add(self, data):
-        """Add a ConjectureData object to the current collection."""
-
-        # First, iterate through the result's buffer, to create the node that
-        # will hold this result. Also note any forced or masked bytes.
-        tree_node = self.nodes[0]
-        indices = []
-        node_index = 0
-        for i, b in enumerate(data.buffer):
-            # We build a list of all the node indices visited on our path
-            # through the tree, since we'll need to refer to them later.
-            indices.append(node_index)
-
-            # If this buffer position was forced or masked, then mark its
-            # corresponding node as forced/masked.
-            if i in data.forced_indices:
-                self.forced[node_index] = b
-            try:
-                self.masks[node_index] = data.masked_indices[i]
-            except KeyError:
-                pass
+        return self.root.is_exhausted
 
-            try:
-                # Use the current byte to find the next node on our path.
-                node_index = tree_node[b]
-            except KeyError:
-                # That node doesn't exist yet, so create it.
-                node_index = len(self.nodes)
-                # Create a new branch node. If this should actually be a leaf
-                # node, it will be overwritten when we store the result.
-                self.nodes.append({})
-                tree_node[b] = node_index
-
-            tree_node = self.nodes[node_index]
-
-            if node_index in self.dead:
-                # This part of the tree has already been marked as dead, so
-                # there's no need to traverse any deeper.
-                break
+    def find_necessary_prefix_for_novelty(self):
+        """Finds a prefix that any novel example must start with.
+        This is currently only used for generate_novel_prefix, where
+        it allows us to significantly speed it up in the case where
+        we start with a very shallow tree.
 
-        # At each node that begins a block, record the size of that block.
-        for b in data.blocks:
-            u, v = b.bounds
-            # This can happen if we hit a dead node when walking the buffer.
-            # In that case we already have this section of the tree mapped.
-            if u >= len(indices):
-                break
-            self.block_sizes[indices[u]] = v - u
-
-        # Forcibly mark all nodes beyond the zero-bound point as dead,
-        # because we don't intend to try any other values there.
-        self.dead.update(indices[self.cap :])
```

Comment:
Is it fine for this to be removed with seemingly no replacement?

(I can imagine it being low-impact; just checking that it's deliberate.)

Reply:
Ah, sorry, I meant to highlight that explicitly: The cap thing is annoying and brittle to include in the tree logic, and ties the caching and generation logic really tightly together in annoying ways. Given that the set of things that it actually helps with there is pretty minimal (essentially it only helps if your buffer size is artifically lowered or your example is incredibly constrained), I decided to leave it out.